<a href="./README.md"><IMG src="./inst/XGR.logo.png" height="100px" id="logo"></a>

#### *XGR: eXploring Genomic Relations at the gene and SNP level through enrichment analysis, similarity analysis and network analysis*

## @ Description
> The central goal of the XGR is to provide a data interpretation system. It is designed to make a user-defined gene or SNP list more interpretable by comprehensively utilising ontology and network information to reveal relationships and enhance opportunities for biological discovery. XGR is unique in supporting a broad range of ontologies (including knowledge of biological and molecular functions, pathways, diseases and phenotypes - in both human and mouse) and different types of networks (including functional, physical and pathway interactions). There are two core functionalities of the XGR. The first is basic infrastructures for easy access to built-in ontologies and networks. The second is extensible functions, now including 1) enrichment analysis using either built-in or custom ontologies, 2) similarity analysis for calculating semantic similarity between genes (or SNPs) based on their ontology annotation profiles, and 3) network analysis for identification of gene subnetworks given a query list of (significant) genes or SNPs. Together with its web app, XGR aims to provide a user-friendly tool for exploring genomic relations at the gene and SNP level.

## @ Installation
> Instructions can be found at [INSTALLATION](./inst/INSTALLATION.md).

## @ Manual
> Manual can be found at [Reference Manual](./inst/HTML.md) and [User Manual](https://rawgit.com/hfang-bristol/XGR/master/vignettes/XGR_vignettes.html).

## @ Citation
> How to cite can be found at [CITATION](./inst/CITATIONS.md).

## @ Web App
> Web-based version of this package can be found at [Web App](http://galahad.well.ox.ac.uk/XGR). 
