<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Function to calculate and visualise correlation</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for xCorrelation"><tr><td>xCorrelation</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Function to calculate and visualise correlation</h2>

<h3>Description</h3>

<p><code>xCorrelation</code> is supposed to calculate and visualise correlation
between a data frame and a named vector (or a list of named vectors).
</p>


<h3>Usage</h3>

<pre>
xCorrelation(df, list_vec, method = c("pearson", "spearman"),
p.type = c("nominal", "empirical"), seed = 825, nperm = 2000,
p.adjust.method = c("BH", "BY", "bonferroni", "holm", "hochberg",
"hommel"),
plot = FALSE, plot.smooth = c(NA, "lm", "loess"))
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>df</code></td>
<td>
<p>a data frame with two columns ('name' and 'value')</p>
</td></tr>
<tr valign="top"><td><code>list_vec</code></td>
<td>
<p>a named vector containing numeric values. Alternatively
it can be a list of named vectors</p>
</td></tr>
<tr valign="top"><td><code>method</code></td>
<td>
<p>the method used to calcualte correlation. It can be
'pearson' for Pearson's correlation or 'spearman' for Spearman rank
correlation</p>
</td></tr>
<tr valign="top"><td><code>p.type</code></td>
<td>
<p>the type of the p-value calcualted. It can be 'nominal'
for nominal p-value or 'empirical' for empirical p-value</p>
</td></tr>
<tr valign="top"><td><code>seed</code></td>
<td>
<p>an integer specifying the seed</p>
</td></tr>
<tr valign="top"><td><code>nperm</code></td>
<td>
<p>the number of random permutations</p>
</td></tr>
<tr valign="top"><td><code>p.adjust.method</code></td>
<td>
<p>the method used to adjust p-values. It can be
one of &quot;BH&quot;, &quot;BY&quot;, &quot;bonferroni&quot;, &quot;holm&quot;, &quot;hochberg&quot; and &quot;hommel&quot;. The
first two methods &quot;BH&quot; (widely used) and &quot;BY&quot; control the false
discovery rate (FDR: the expected proportion of false discoveries
amongst the rejected hypotheses); the last four methods &quot;bonferroni&quot;,
&quot;holm&quot;, &quot;hochberg&quot; and &quot;hommel&quot; are designed to give strong control of
the family-wise error rate (FWER). Notes: FDR is a less stringent
condition than FWER</p>
</td></tr>
<tr valign="top"><td><code>plot</code></td>
<td>
<p>logical to indicate whether scatter plot is drawn</p>
</td></tr>
<tr valign="top"><td><code>plot.smooth</code></td>
<td>
<p>the smooth method for the scatter plot. It can be NA
(depending on correlation type), &quot;lm&quot; for the linear line or 'loess'
for the loess curve</p>
</td></tr>
</table>


<h3>Value</h3>

<p>a list with three componets:
</p>

<ul>
<li><p><code>df_summary</code>: a data frame of n x 5, where n is the number
of named vectors, and the 5 columns are &quot;name&quot;, &quot;num&quot; (i.e. number of
data points used for calculation), &quot;cor&quot; (i.e. correlation), &quot;pval&quot;
(i.e. p-value), &quot;fdr&quot;
</p>
</li>
<li><p><code>ls_gp_curve</code>: NULL if the plot is not drawn; otherwise, a
list of 'ggplot' objects for scatter plot together with an estimated
curve
</p>
</li>
<li><p><code>ls_gp_pdf</code>: NULL if the plot is not drawn; otherwise, a
list of 'ggplot' objects for pdf plot for null distribution of
correlation together with a vertical line for observed correlation
</p>
</li></ul>



<h3>Note</h3>

<p>none
</p>


<h3>See Also</h3>

<p><code>xCorrelation</code>
</p>


<h3>Examples</h3>

<pre>
## Not run: 
# Load the library
library(XGR)

## End(Not run)

RData.location &lt;- "http://galahad.well.ox.ac.uk/bigdata"
## Not run: 
# a) provide the seed nodes/genes with the weight info
## load ImmunoBase
ImmunoBase &lt;- xRDataLoader(RData.customised='ImmunoBase',
RData.location=RData.location)
## get genes within 500kb away from AS GWAS lead SNPs
seeds.genes &lt;- ImmunoBase$AS$genes_variants
## seeds weighted according to distance away from lead SNPs
data &lt;- 1- seeds.genes/500000

# b) prepare a data frame
df &lt;- data.frame(name=names(data), value=data, stringsAsFactors=FALSE)

# c) do correlation
ls_res &lt;- xCorrelation(df, data, method="pearson", p.type="empirical",
nperm=2000, plot=TRUE)

## End(Not run)
</pre>


</body></html>
