<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Function to visualise prioritised genes using manhattan plot</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for xPrioritiserManhattan"><tr><td>xPrioritiserManhattan</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Function to visualise prioritised genes using manhattan plot</h2>

<h3>Description</h3>

<p><code>xPrioritiserManhattan</code> is supposed to visualise prioritised genes
using manhattan plot. Genes with the top priority are highlighed. It
returns an object of class &quot;ggplot&quot;.
</p>


<h3>Usage</h3>

<pre>
xPrioritiserManhattan(pNode, color = c("darkred", "darkgreen"), cex =
0.5,
highlight.top = 20, highlight.col = "deepskyblue",
highlight.label.size = 2, highlight.label.offset = 0.02,
highlight.label.col = "darkblue", verbose = T,
RData.location =
"https://github.com/hfang-bristol/RDataCentre/blob/master/XGR/1.0.0")
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>pNode</code></td>
<td>
<p>an object of class &quot;pNode&quot;</p>
</td></tr>
<tr valign="top"><td><code>color</code></td>
<td>
<p>a character vector for point colors to alternate</p>
</td></tr>
<tr valign="top"><td><code>cex</code></td>
<td>
<p>a numeric value for point size</p>
</td></tr>
<tr valign="top"><td><code>highlight.top</code></td>
<td>
<p>the number of the top targets to be highlighted</p>
</td></tr>
<tr valign="top"><td><code>highlight.col</code></td>
<td>
<p>the highlight colors</p>
</td></tr>
<tr valign="top"><td><code>highlight.label.size</code></td>
<td>
<p>the highlight label size</p>
</td></tr>
<tr valign="top"><td><code>highlight.label.offset</code></td>
<td>
<p>the highlight label offset</p>
</td></tr>
<tr valign="top"><td><code>highlight.label.col</code></td>
<td>
<p>the highlight label color</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>logical to indicate whether the messages will be
displayed in the screen. By default, it sets to false for no display</p>
</td></tr>
<tr valign="top"><td><code>RData.location</code></td>
<td>
<p>the characters to tell the location of built-in
RData files. See <code>xRDataLoader</code> for details</p>
</td></tr>
</table>


<h3>Value</h3>

<p>an object of class &quot;ggplot&quot;
</p>


<h3>Note</h3>

<p>none
</p>


<h3>See Also</h3>

<p><code>xRDataLoader</code>, <code>xPrioritiser</code>,
<code>xPrioritiserSNPs</code>, <code>xPrioritiserGenes</code>,
<code>xPrioritiserPathways</code>
</p>


<h3>Examples</h3>

<pre>
## Not run: 
# Load the library
library(XGR)
library(igraph)
library(dnet)
library(ggbio)

# a) provide the SNPs with the significance info
## get lead SNPs reported in AS GWAS and their significance info (p-values)
AS &lt;- read.delim(file.path(path.package("XGR"),"AS.txt"),
stringsAsFactors=FALSE)

# b) perform priority analysis
pNode &lt;- xPrioritiserSNPs(data=AS,
network="PCommonsUN_medium",restart=0.7)

# c) manhattan plot
mp &lt;- xPrioritiserManhattan(pNode, highlight.top=10)
#pdf(file="Gene_manhattan.pdf", height=6, width=12, compress=TRUE)
print(mp)
#dev.off()

## End(Not run)
</pre>


</body></html>
