<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Function to extract genomic locations given a list of SNPs</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for xSNPlocations"><tr><td>xSNPlocations</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Function to extract genomic locations given a list of SNPs</h2>

<h3>Description</h3>

<p><code>xSNPlocations</code> is supposed to extract genomic locations given a
list of SNPs.
</p>


<h3>Usage</h3>

<pre>
xSNPlocations(
data,
GR.SNP = c("dbSNP_GWAS", "dbSNP_Common", "dbSNP_Single"),
verbose = T,
RData.location = "http://galahad.well.ox.ac.uk/bigdata",
guid = NULL
)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>data</code></td>
<td>
<p>a input vector containing SNPs. SNPs should be provided as
dbSNP ID (ie starting with rs). Alternatively, they can be in the
format of 'chrN:xxx', where N is either 1-22 or X, xxx is genomic
positional number; for example, 'chr16:28525386'</p>
</td></tr>
<tr valign="top"><td><code>GR.SNP</code></td>
<td>
<p>the genomic regions of SNPs. By default, it is
'dbSNP_GWAS', that is, SNPs from dbSNP (version 146) restricted to GWAS
SNPs and their LD SNPs (hg19). It can be 'dbSNP_Common', that is,
Common SNPs from dbSNP (version 146) plus GWAS SNPs and their LD SNPs
(hg19). Alternatively, the user can specify the customised input. To do
so, first save your RData file (containing an GR object) into your
local computer, and make sure the GR object content names refer to
dbSNP IDs. Then, tell &quot;GR.SNP&quot; with your RData file name (with or
without extension), plus specify your file RData path in
&quot;RData.location&quot;. Note: you can also load your customised GR object
directly</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>logical to indicate whether the messages will be
displayed in the screen. By default, it sets to true for display</p>
</td></tr>
<tr valign="top"><td><code>RData.location</code></td>
<td>
<p>the characters to tell the location of built-in
RData files. See <code>xRDataLoader</code> for details</p>
</td></tr>
<tr valign="top"><td><code>guid</code></td>
<td>
<p>a valid (5-character) Global Unique IDentifier for an OSF
project. See <code>xRDataLoader</code> for details</p>
</td></tr>
</table>


<h3>Value</h3>

<p>an GR oject, with an additional metadata column called 'variant_id'
storing SNP location in the format of 'chrN:xxx', where N is either
1-22 or X, xxx is genomic positional number.
</p>


<h3>Note</h3>

<p>none
</p>


<h3>See Also</h3>

<p><code>xRDataLoader</code>
</p>


<h3>Examples</h3>

<pre>
## Not run: 
# Load the XGR package and specify the location of built-in data
library(XGR)

## End(Not run)
RData.location &lt;- "http://galahad.well.ox.ac.uk/bigdata"

## Not run: 
# a) provide the SNPs
## load ImmunoBase
ImmunoBase &lt;- xRDataLoader(RData.customised='ImmunoBase',
RData.location=RData.location)
## get lead SNPs reported in AS GWAS and their significance info (p-values)
gr &lt;- ImmunoBase$AS$variant
data &lt;- names(gr)

# b) find the location
snp_gr &lt;- xSNPlocations(data=data, RData.location=RData.location)

## End(Not run)
</pre>


</body></html>
