<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Function to define eQTL genes given a list of SNPs or a...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for xSNP2eGenes"><tr><td>xSNP2eGenes</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Function to define eQTL genes given a list of SNPs or a customised eQTL mapping data</h2>

<h3>Description</h3>

<p><code>xSNP2eGenes</code> is supposed to define eQTL genes given a list of
SNPs or a customised eQTL mapping data. The eQTL weight is calcualted
as Cumulative Distribution Function of negative log-transformed
eQTL-reported signficance level.
</p>


<h3>Usage</h3>

<pre>
xSNP2eGenes(data, include.eQTL = NA, eQTL.customised = NULL,
cdf.function = c("empirical", "exponential"), plot = FALSE,
verbose = TRUE,
RData.location = "http://galahad.well.ox.ac.uk/bigdata")
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>data</code></td>
<td>
<p>an input vector containing SNPs. SNPs should be provided as
dbSNP ID (ie starting with rs). Alternatively, they can be in the
format of 'chrN:xxx', where N is either 1-22 or X, xxx is number; for
example, 'chr16:28525386'</p>
</td></tr>
<tr valign="top"><td><code>include.eQTL</code></td>
<td>
<p>the eQTL supported currently. By default, it is
'NA' to disable this option. Pre-built eQTL datasets are detailed in
<code>xDefineEQTL</code></p>
</td></tr>
<tr valign="top"><td><code>eQTL.customised</code></td>
<td>
<p>a user-input matrix or data frame with 4
columns: 1st column for SNPs/eQTLs, 2nd column for Genes, 3rd for eQTL
mapping significance level (p-values or FDR), and 4th for contexts
(required even though only one context is input). Alternatively, it can
be a file containing these 4 columns. It is designed to allow the user
analysing their eQTL data. This customisation (if provided) will
populate built-in eQTL data</p>
</td></tr>
<tr valign="top"><td><code>cdf.function</code></td>
<td>
<p>a character specifying a Cumulative Distribution
Function (cdf). It can be one of 'exponential' based on exponential
cdf, 'empirical' for empirical cdf</p>
</td></tr>
<tr valign="top"><td><code>plot</code></td>
<td>
<p>logical to indicate whether the histogram plot (plus
density or CDF plot) should be drawn. By default, it sets to false for
no plotting</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>logical to indicate whether the messages will be
displayed in the screen. By default, it sets to true for display</p>
</td></tr>
<tr valign="top"><td><code>RData.location</code></td>
<td>
<p>the characters to tell the location of built-in
RData files. See <code>xRDataLoader</code> for details</p>
</td></tr>
</table>


<h3>Value</h3>

<p>a data frame with following columns:
</p>

<ul>
<li><p><code>Gene</code>: eQTL-containing genes
</p>
</li>
<li><p><code>SNP</code>: eQTLs
</p>
</li>
<li><p><code>Sig</code>: the eQTL mapping significant level (the
best/minimum)
</p>
</li>
<li><p><code>Weight</code>: the eQTL weight
</p>
</li></ul>



<h3>Note</h3>

<p>none
</p>


<h3>See Also</h3>

<p><code>xRDataLoader</code>
</p>


<h3>Examples</h3>

<pre>
## Not run: 
# Load the library
library(XGR)

## End(Not run)

RData.location &lt;- "http://galahad.well.ox.ac.uk/bigdata"
## Not run: 
# a) provide the SNPs with the significance info
data(ImmunoBase)
gr &lt;- ImmunoBase$AS$variants
AS &lt;- as.data.frame(GenomicRanges::mcols(gr)[, c('Variant','Pvalue')])

# b) define eQTL genes
df_eGenes &lt;- xSNP2eGenes(data=AS[,1], include.eQTL="JKscience_TS2A",
RData.location=RData.location)

## End(Not run)
</pre>


</body></html>
